src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,169 @@
-
+        if (index < 0 || index > currentSlot + 1) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds ");
+        }
+        checkAndIncreaseArrayLength();
+        currentSlot++;
+        for (int i = size() - 1; i >= (index == 0 ? index + 1 : index); i--) {
+            array[i] = array[i - 1];
+        }
+        array[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        checkAndIncreaseArrayLength(list.size());
+        for (int i = 0; i < list.size(); i++) {
+            currentSlot++;
+            array[currentSlot] = list.get(i);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index > currentSlot) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds ");
+        }
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index > currentSlot) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds ");
+        }
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size()) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds");
+        }
+        T date = array[index];
+        for (int i = index; i < size(); i++) {
+            if (index == currentSlot) {
+                array[index] = null;
+                currentSlot--;
+                return date;
+            }
+            array[i] = array[i + 1];
+        }
+        currentSlot--;
+        array[size()] = null;
+        return date;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        if (isEmpty()) {
+            throw new NoSuchElementException("Found element doesn't exist");
+        }
+        int indexFoundElement = -1;
+        search:
+        for (int i = 0; i < size(); i++) {
+            if (array[i] == null || element == null) {
+                indexFoundElement = i;
+                break search;
+            }
+            if (array[i] == element || array[i].equals(element)) {
+                indexFoundElement = i;
+                break search;
+            }
+        }
+        if (indexFoundElement == -1) {
+            throw new NoSuchElementException("Found element doesn't exist");
+        } else {
+            for (int i = indexFoundElement; i < size(); i++) {
+                if (indexFoundElement == size() - 1) {
+                    array[indexFoundElement] = null;
+                    currentSlot--;
+                    return element;
+                }
+                array[i] = array[i + 1];
+            }
+            array[currentSlot] = null;
+            currentSlot--;
+        }
+        return element;
```

Comment:
same here, just null check every element up to the `size`, and if you found it - use `arraycopy()` to move the elements that are behind it